### DISCUSSION TITLE
Directus Loader for Astro Content Layer

### DISCUSSION THREAD

**Discussion by rickylabs** [directus/directus#23771] [Feature Requests]
# Proposal: Developing a Directus Loader for Astro Content Layer

## Context

Astro's Content Layer API allows developers to load content from various sources. We propose developing a Directus Loader for Astro, similar to existing loaders for Strapi, Notion, and Airtable. This would enable Astro developers to easily integrate Directus content into their projects.

[Astro Blog - Content Layer: A Deep Dive](https://astro.build/blog/content-layer-deep-dive/)

## Basic Example

```typescript
import { defineCollection } from 'astro:content';
import { directusLoader } from 'astro-directus-loader';

const posts = defineCollection({
  loader: directusLoader({
    apiUrl: 'https://your-directus-instance.com',
    collection: 'posts',
    // Optional: authentication token
    token: process.env.DIRECTUS_TOKEN
  }),
  schema: z.object({
    title: z.string(),
    content: z.string(),
    author: z.string(),
    publishDate: z.date()
  })
});

// Use in Astro components
const allPosts = await getCollection('posts');
const specificPost = await getEntry('posts', '1');
```
### Astro Loader official implementations:
- [StoryBlok](https://github.com/storyblok/storyblok-astro/blob/alpha/lib/content-layer/storyblokLoader.ts)
- [Hygraph ](https://github.com/hygraph/hygraph-astro-loader/blob/main/src/HygraphLoader.ts)
- [Cloudinary](https://github.com/cloudinary-community/astro-cloudinary/blob/main/astro-cloudinary/src/loaders/assets-loader.ts)
### Astro Loader community implementations examples:
- [Strapi - detailled article](https://strapi.io/blog/how-to-create-a-custom-astro-loader-for-strapi-using-content-layer-api)
- [Notion](https://github.com/NotWoods/travel/tree/main/packages/notion-astro-loader#readme)
- [Airtable](https://github.com/ascorbic/astro-loaders/tree/main/packages/airtable)

## Motivation

1. **Astro Integration**: Allow Astro developers to easily use Directus as a content source.
2. **Type Safety**: Leverage Astro's type-safe content approach with Directus data.
3. **Performance**: Utilize Astro's efficient content caching for Directus content.
4. **Ecosystem Growth**: Increase Directus' presence in the Astro ecosystem.

### Directus Content Layer API: 
https://5-0-0-beta.docs.astro.build/en/reference/loader-reference/

1. **Loader Function**: Create a `directusLoader` function that connects to Directus API.
2. **Data Fetching**: Implement methods to fetch collections and items from Directus.
3. **Caching**: Utilize Astro's built-in caching mechanism for fetched Directus content allowing incremental updates.
4. **Schema Mapping**: Convert Directus field types to appropriate Zod schemas.

## Requirements

Must Have:
- Basic authentication support
- Fetch collections and items from Directus
- Map Directus fields to Zod schemas
- Error handling and logging

Should Have:
- Support for Directus filters and sorting
- Pagination for large datasets
- Support for Directus relational fields

Could Have:
- Support for Directus file fields
- Real-time updates (if Astro supports it)

> ### What’s next
> The Content Layer API is a big step forward for Astro, but we’re just getting started. Currently, the data store is just a key-value store, with limited filtering. This is fast but not memory efficient, and querying is not very flexible. Our goal is to introduce a backend based on Astro DB in a future version to help it scale to hundreds of thousands of pages. It will also allow us to support more complex queries, and perhaps even real-time updates.
> _source: [Astro Blog](https://astro.build/blog/content-layer-deep-dive/)_

Won't Have (initially):
- Direct editing of Directus content through Astro

## Implementation Steps

1. Set up the loader package structure
2. Implement basic Directus API connection
3. Create data fetching and caching logic
4. Develop schema mapping functionality
5. Add authentication and error handling
6. Write documentation and examples

## Unresolved Questions

1. How to handle Directus-specific features like roles and permissions?
2. What's the best way to map complex Directus field types to Zod schemas?
3. How to optimize performance for very large Directus collections?

**Comment by somethingxdev** [comment]
+1

**Comment by pelleknaap** [comment]
This would be an amazing addition to Directus and make the development process of Directus & Astro rock solid. 
The current development experience is subpar due to the whole type situation. There simply isn't a good solution to automatically pull types from the CMS and use them in Astro. The current codegen solutions make all fields optional (which is a hard issue to solve) and manually writing the types is error-prone and defeats a large portion of type safety.

Creating a Directus loader for Astro content collections would simply solve the whole issue by providing a solid API to pull data, coupled with type safety provided by Zod. It would be amazing if we can get this implemented and it would put Directus on the map for a lot of Astro developers.

**Reply by pelleknaap** [reply]
The Astro part of [this Strapi blog post](https://strapi.io/blog/how-to-create-a-custom-astro-loader-for-strapi-using-content-layer-api) explains the new content-layer well and gives a good example implementation for a similar CMS. I will try to create a rough prototype soon if I can find the time.

**Reply by somethingxdev** [reply]
> The Astro part of [this Strapi blog post](https://strapi.io/blog/how-to-create-a-custom-astro-loader-for-strapi-using-content-layer-api) explains the new content-layer well and gives a good example implementation for a similar CMS. I will try to create a rough prototype soon if I can find the time.

How's it going?

**Comment by Jonnotie** [comment]
Yes please!

**Comment by rxlxnd** [comment]
+2

**Comment by jjaimealeman** [comment]
Another vote for Directus Loader for Astro Content Layer 👍🏻

**Comment by kasvith** [comment]
+1

**Comment by rijkvanzanten** [comment]
Good shout 🙂 I know absolutely nothing about Astro or how it's data loading works though... 

From a quick read through some of the links; am I correctly understanding that we'd effectively provide a translations layer that implement the Astro calls and "just" calls the Directus SDK from there? The [Hygraph loader source](https://github.com/hygraph/hygraph-astro-loader/blob/main/src/HygraphLoader.ts) looks like a very basic "Astro calls `fetchAllData`; this loader just calls the GraphQL endpoint" type of setup.

@rickylabs @pelleknaap

**Reply by rijkvanzanten** [reply]
Quick PS as it's happened a lot on this thread: a `+1` comment is very unhelpful and just adds noise 😬 

It pings everybody on this thread with no new information, and there's no way to actually sort by number of +1 comments in the list of feature requests. Please just upvote the original post or leave an emoji reaction there instead 🙏🏻

**Reply by somethingxdev** [reply]
any news?

**Reply by rickylabs** [reply]
@rijkvanzanten Thanks for the response! Yes, you're correct about providing a translation layer between Astro and Directus SDK. However, there are several key aspects we need to consider:

## Type Safety

1. **Input Types (Directus Schema)**
Given the ongoing work on [TypeScript type generation from Directus schema](https://github.com/directus/directus/issues/18121), we could leverage this to ensure full type safety without manual type definitions. The loader would look something like:

```typescript
import { createDirectus, rest } from '@directus/sdk';
// These types would be auto-generated from Directus schema
import type { Schema } from './directus-types';

export function directusLoader(options: {
  apiUrl: string;
  collection: keyof Schema;
  token?: string;
}) {
  const client = createDirectus<Schema>(options.apiUrl).with(rest());
  
  return async ({ store, meta }) => {  // Using Astro's loader context
    const lastModified = meta.get('lastModified');
    
    const response = await client.request(
      readItems(options.collection)
    );
    
    // Transform response for Astro Content Layer
    return response.map(item => ({
      id: String(item.id),  // Ensure string ID for Astro
      ...item
    }));
  };
}
```

ideally the Schema would be generated by CLI command

2. **Output Types (Astro Content Layer)**
For the Astro side, we could infer the Zod schema from the Directus types:

```typescript
type CollectionType = Awaited<ReturnType<typeof client.items<'collection_name'>.readByQuery>>;

// This could be a utility to generate Zod schemas from Directus types
const zodSchemaFromDirectus = createZodSchema<CollectionType>();

const collection = defineCollection({
  loader: directusLoader({
    apiUrl: 'https://directus.example.com',
    collection: 'collection_name'
  }),
  schema: zodSchemaFromDirectus
});
```

Or we could leverage a solution like this one it allows local schema definition override:
https://github.com/pawcoding/astro-loader-pocketbase/blob/master/src/generate-schema.ts
but by default generate schema from remote source:
https://github.com/pawcoding/astro-loader-pocketbase/blob/master/src/utils/get-remote-schema.ts

We could do this in Directus using GraphQL and OAS endpoints and convert them to schema using a cli command like https://github.com/astahmer/openapi-zod-client#openapi-zod-client or https://github.com/withshepherd/graphql-codegen-zod

## Caching & Content Updates

3. **Efficient Caching**
We should leverage Astro's Content Layer caching capabilities:
```typescript
export function directusLoader(options: {/* ... */}) {
  return async ({ store, meta, logger }) => {
    const lastModified = meta.get('lastModified');
    const digest = meta.get('contentDigest');

    // Check if content has changed using Directus' revision system
    const hasChanged = await checkContentChanges(client, lastModified);
    
    if (!hasChanged) {
      logger.info('Content unchanged, using cached version');
      return;
    }

    // Fetch and update content...
    meta.set('lastModified', new Date().toISOString());
    meta.set('contentDigest', generateDigest(newContent));
  };
}
```
More detailled implementation example here:  
https://github.com/pawcoding/astro-loader-pocketbase/blob/master/src/pocketbase-loader.ts#L60
and here:
https://github.com/pawcoding/astro-loader-pocketbase/blob/master/src/load-entries.ts

4. **Draft Mode & Live Revalidation**
We need to consider:
- How to handle draft content in development
- Live revalidation mechanism for content updates
- Integration with Directus' webhook system for real-time updates

Here is a simplified approach using Directus Flows to trigger webhook on data source change :

```typescript
export default function directusWebhook() {
  return {
    name: "directus-webhook",
    hooks: {
      "astro:server:setup": async ({ server, refreshContent }) => {
        // Register a dev server webhook endpoint
        server.middlewares.use("/_refresh", async (req, res) => {
          if (req.method !== "POST") {
            res.statusCode = 405;
            res.end("Method Not Allowed");

            return;
          }

          let body = "";

          req.on("data", (chunk) => {
            body += chunk.toString();
          });
          req.on("end", async () => {
            try {
              const webhookBody = JSON.parse(body);

              console.log("Webhook body:", webhookBody);
              await refreshContent({
                context: { webhookBody },
                loaders: ["directus-pages-loader"],
              });
              res.writeHead(200, { "Content-Type": "application/json" });
              res.end(
                JSON.stringify({ message: "Content refreshed successfully" }),
              );
            } catch (error) {
              res.writeHead(500, { "Content-Type": "application/json" });
              res.end(
                JSON.stringify({
                  error: "Failed to refresh content: " + error.message,
                }),
              );
            }
          });
        });
      },
    },
  };
}
```
```Typescript
export function pagesLoader(): Loader {
  return {
    name: "directus-pages-loader",
    load: async ({
      store,
      logger,
      parseData,
      meta,
      generateDigest,
      refreshContextData,
    }) => {
      const apiUrl = "http://localhost:8055";
      const collection = "pages";
      const url = `${apiUrl}/items/${collection}`;

      if (refreshContextData) {
        logger.info(`Refreshing context data... `);
        console.log(refreshContextData);
        //ToDo - handle partial refresh
      }
```
Questions for the team:
1. Once type generation is implemented, could we expose utilities to help **generate Zod schemas from Directus types**?
2. Would it make sense to provide a standard way to transform Directus field types to Zod schemas?
3. How should we handle Directus-specific features like **field permissions** and **nested collections** in the type generation?
4. What's the recommended approach for **tracking content changes in Directus for efficient caching**?
5. Would it be possible to expose a standardized way to listen for content updates that could be used with Astro's development server ? maybe a **Directus extension**?

I can help work on this integration once we have a clear path forward for:
- Type generation
- Content change detection for caching
- Draft/preview mode implementation

**Reply by rijkvanzanten** [reply]
@rickylabs Would you say the typescript generation endpoint is a _requirement_ for this to be a successful content layer, or a nice to have? Sounds like theoretically it could work without but the experience wouldn't nearly be as nice as it could/should? be

**Reply by rickylabs** [reply]
@rijkvanzanten In my opinion, type generation is a requirement. While loading Directus data in Content Layer can be done without an official loader as you pointed out, the main purpose of an official Loader would be to handle more complex implementations: caching, live revalidation, preview (draft mode), and most importantly, type safety - which is the big selling point of Content Layer.

We could imagine a first version without typegen, which might already be interesting for some folks doing side projects. However, as soon as you're a company or individual building a serious website, blog, or SaaS that could potentially handle hundreds of pages and thousands of fields across many collections, this kind of feature becomes a decisive factor when choosing your tech stack.

There is an ongoing discussion on Astro Discord [[Mega Thread] - Best CMS for Astro](https://discord.com/channels/830184174198718474/1222578561261637734) where Directus is mentioned frequently as the perfect match for Astro. One of the main reasons is its compatibility with SQLite DB (Turso), which Astro is pushing hard with its own service, Astro DB.

I also believe that with a proper content layer and collaborative effort from both teams for smart integration between the two systems, it could be an excellent candidate as a WordPress replacement at a time when many people are seeking a solution to move away from WordPress.

Once we have a proper Content Layer and auto typegen, we could even build a Directus starter with common web blocks - a Directus extension paired with an Astro starter that handles everything using solutions like route groups and dynamic segments, similar to what we see in:
- Next.js with Builder.io
- Next.js with Plasmic

Once we get a proper Directus Content Layer working, I'd be happy to work on the latter if anyone's interested.

**Reply by rijkvanzanten** [reply]
Thanks for all the insights @rickylabs! Really helps me understand what the scope would be here 🙂 

Last quick technical question now that I have you: does Astro differentiate between mutations and querying in their type signatures, or is it all one schema you're working with? Mostly making sure that we account for that difference in the typegen endpoint if it's relevant

**Reply by rickylabs** [reply]
@rijkvanzanten Short answer from what I understand of Content Layer's purpose: No, it doesn't.

As stated in their blog post:
> The important thing to note here is that the data store is only updated at build time: a deployed site cannot change the data store.
Source: https://astro.build/blog/content-layer-deep-dive/

For better understanding, see this Astro build lifecycle illustration:
![17335237626417415887205261527373](https://github.com/user-attachments/assets/6c217e16-3b4e-49bc-9cc2-d915e3bf4412)

From my understanding, Astro doesn't plan to handle data mutation with Content Layer for now as it's a completely different data lifecycle, and therefore you cannot guarantee that data structure remains valid. It would be extremely challenging to handle both build-time and runtime type safety using the same solution.

Perhaps @matthewp could shed some light on what would be the most idiomatic way to handle type safety in both build-time and runtime lifecycles, and if Content Layer will someday handle both.

**Reply by rickylabs** [reply]
@rijkvanzanten in addition to that, could you confirm/answer my original questions?

> 1. Once type generation is implemented, could we expose utilities to help generate Zod schemas from Directus types?

What would be really awesome would be to also offer the possibility to generate Zod schemas from CLI - in fact, in Astro, it's even more important than typegen itself.

Behind the scenes, we could use [ts-to-zod](https://github.com/fabien0102/ts-to-zod) and "fix" complex Directus types to return an "almost" perfect schema.

From there, people could freely import it in Content Layer outputs and extend it if needed.

> 2. Would it make sense to provide a standard way to transform Directus field types to Zod schemas?

If #1 could not be provided, we should at least provide some basic helpers for Directus relationship models and more complex fields. Additionally, perhaps we could provide another utility to infer Zod schema depending on access policy - what do you think?

> 3. How should we handle Directus-specific features like field permissions and nested collections in the type generation?

?

> 4. What's the recommended approach for tracking content changes in Directus for efficient caching?

?

> 5. Would it be possible to expose a standardized way to listen for content updates that could be used with Astro's development server? Maybe a Directus extension?

We could leverage Directus websocket connection to listen for collection changes on demand in Dev mode.

Thanks in advance for your insights ! Eager to see what we can do here !

**Reply by rijkvanzanten** [reply]
1. I find this one tricky to defend for API-driven codegen, as zod — while popular — not a standard or spec. Once we ship that, it'll be T-5 seconds before people expect/request support for typebox/valibot/joi/ajv/yup etc. That being said, if there's a minimal external lib that'll gen zod from "directus flavored types" that could work 🤔 
2. I don't really know enough about zod to answer this.. That being said, sounds like this could "easily" be a small CLI tool that uses the Directus APIs to codegen some Zod JS with these niceties included 🙂 
3. Field permissions should be based on the token that the API is queried with — that's how we've done it for GQL / OAS previously. Nested collections in GQL/OAS are just that; nested references. In our SDK we allow/expect a union type for the primary key type | nested object in TS. I'm thinking we should align the types generation with our own SDK first as it's the most widely supported integration. I'll have to do a bit more research to see if that'll align with what Astro expects or if that'll blow up spectacularly.
4. I'd say the simplest thing is to add a `last_modified` field that's set to auto-save the current timestamp every save 👍🏻 
5. Yep, websockets! Either GQL subscriptions or our websockets would be ideal for that 🙂 

The more I'm thinking about this, the more I'm feeling the codegen side here could/should? be a separate CLI altogether instead of baking it into our `/specs` endpoint. That means we can be a bit more specific in what flavors we support, add a bunch more opt-in options like "Make it Astro flavored", and have it generate multiple files which would be huge for larger setups 👀

**Reply by somethingxdev** [reply]
> 1. I find this one tricky to defend for API-driven codegen, as zod — while popular — not a standard or spec. Once we ship that, it'll be T-5 seconds before people expect/request support for typebox/valibot/joi/ajv/yup etc. That being said, if there's a minimal external lib that'll gen zod from "directus flavored types" that could work 🤔
> 
>     2. I don't really know enough about zod to answer this.. That being said, sounds like this could "easily" be a small CLI tool that uses the Directus APIs to codegen some Zod JS with these niceties included 🙂
> 
>     3. Field permissions should be based on the token that the API is queried with — that's how we've done it for GQL / OAS previously. Nested collections in GQL/OAS are just that; nested references. In our SDK we allow/expect a union type for the primary key type | nested object in TS. I'm thinking we should align the types generation with our own SDK first as it's the most widely supported integration. I'll have to do a bit more research to see if that'll align with what Astro expects or if that'll blow up spectacularly.
> 
>     4. I'd say the simplest thing is to add a `last_modified` field that's set to auto-save the current timestamp every save 👍🏻
> 
>     5. Yep, websockets! Either GQL subscriptions or our websockets would be ideal for that 🙂
> 
> 
> The more I'm thinking about this, the more I'm feeling the codegen side here could/should? be a separate CLI altogether instead of baking it into our `/specs` endpoint. That means we can be a bit more specific in what flavors we support, add a bunch more opt-in options like "Make it Astro flavored", and have it generate multiple files which would be huge for larger setups 👀

Hi. Any news?

**Reply by zachinglis** [reply]
> Would it be possible to expose a standardized way to listen for content updates that could be used with Astro's development server? Maybe a Directus extension?

I'd argue that Git hooks to redeploy on content change would be a better solution for most people. Especially as git deployed hosting solutions like Vercel use caches for most of the content. 

It'd be good if it was debounced though.

**Reply by rickylabs** [reply]
> 1. I find this one tricky to defend for API-driven codegen, as zod — while popular — not a standard or spec. Once we ship that, it'll be T-5 seconds before people expect/request support for typebox/valibot/joi/ajv/yup etc. That being said, if there's a minimal external lib that'll gen zod from "directus flavored types" that could work 🤔
> 2. I don't really know enough about zod to answer this.. That being said, sounds like this could "easily" be a small CLI tool that uses the Directus APIs to codegen some Zod JS with these niceties included 🙂
> 3. Field permissions should be based on the token that the API is queried with — that's how we've done it for GQL / OAS previously. Nested collections in GQL/OAS are just that; nested references. In our SDK we allow/expect a union type for the primary key type | nested object in TS. I'm thinking we should align the types generation with our own SDK first as it's the most widely supported integration. I'll have to do a bit more research to see if that'll align with what Astro expects or if that'll blow up spectacularly.
> 4. I'd say the simplest thing is to add a `last_modified` field that's set to auto-save the current timestamp every save 👍🏻
> 5. Yep, websockets! Either GQL subscriptions or our websockets would be ideal for that 🙂
> 
> The more I'm thinking about this, the more I'm feeling the codegen side here could/should? be a separate CLI altogether instead of baking it into our `/specs` endpoint. That means we can be a bit more specific in what flavors we support, add a bunch more opt-in options like "Make it Astro flavored", and have it generate multiple files which would be huge for larger setups 👀

https://standardschema.dev/ @rijkvanzanten  this is the way ! 😉

**Comment by Jihad** [comment]
Just coming into both Astro and Directus, one of the great features of astro is how simple and flexible it is. The many benefits of  [content collections](https://docs.astro.build/en/guides/content-collections/) are really attrative.

Any one found a way ot workaround for directus loader? Saw lots of [community loaders](https://astro.build/integrations/?search=&categories%5B%5D=loaders) for different sources but couldn't find one for directus. And tbh I lack the skills to make one.

**Reply by marekbrze** [reply]
> Just coming into both Astro and Directus, one of the great features of astro is how simple and flexible it is. The many benefits of  [content collections](https://docs.astro.build/en/guides/content-collections/) are really attrative.
> 
> 
> 
> Any one found a way ot workaround for directus loader? Saw lots of [community loaders](https://astro.build/integrations/?search=&categories%5B%5D=loaders) for different sources but couldn't find one for directus. And tbh I lack the skills to make one.

Here is the simple example how to combine Directus SDK with static Content Loader 

```typescript 
const cities = defineCollection({
  loader: async () => {
    const request = await directus.request(
      readItems("Miasta", {
        fields: ["id", "Name", "Slug", { Wojewodztwo: ["Slug"] }],
        limit: -1,
      }),
    );
    return request.map((city) => ({
      id: city.Slug,
      intId: city.id,
      name: city.Name,
      region: city.Wojewodztwo.Slug,
    }));
  },
  schema: z.object({
    id: z.string(),
    name: z.string(),
    region: z.string(),
  }),
});
```

**Reply by Jihad** [reply]
@marekbrze Thanks! awesome start to go from there.

**Reply by marekbrze** [reply]
> @marekbrze Thanks! awesome start to go from there.

Basically you have 3 sections:
1. Directus request (read about their JS SDK
2. Structuring the data from Directus for Astro - I learned that Astro requires "string" ids, not "number" ids, so in my code I pass slug as an id
3. Zod structure of the collection

**Reply by Jihad** [reply]
> > @marekbrze Thanks! awesome start to go from there.
> 
> Basically you have 3 sections:
> 
> 1. Directus request (read about their JS SDK
> 2. Structuring the data from Directus for Astro - I learned that Astro requires "string" ids, not "number" ids, so in my code I pass slug as an id
> 3. Zod structure of the collection

Great! Yeah noticed that too. Also, apparently, i'll need to use a custom markdown parser - can't use Astro built-in feature or customization defaults-. That requires a custom object loader not inline. I guess it'll work fine, found couple custom md parsers.

**Reply by Cryszon** [reply]
You can manually configure the same Markdown parser Astro uses* and add any additional plugins you want. That's what I did.

> [!WARNING]
> 
> The code snippet below is vulnerable to XSS as it allows unsafe HTML. Check out the full source with sanitization [here](https://github.com/ReunMedia/company-website/blob/d0b4d78692212ae7e8194ce10cb96583887a9363/apps/frontend-astro/src/utils/markdownToHtml.ts).

```ts
const rootProcessor = unified()
  .use(remarkParse) // Base Markdown parser
  // Astro includes `remark-gfm` and `remark-smartypants` by default so we
  // include them to mimick that.
  .use(remarkGfm)
  .use(remarkSmartypants)
  .use(remarkRehype, { allowDangerousHtml: true } satisfies RemarkRehypeOptions)
  .use(rehypeSlug) // Add IDs to headings
  .use(rehypeRaw)
  // Code block syntax highlight
  .use(rehypeShiki, {
    themes: {
      light: "github-light",
      dark: "github-dark",
    },
    colorReplacements: {
      "github-light": {
        "#fff": "var(--color-theme-background-highlight)",
      },
      "github-dark": {
        "#24292e": "var(--color-theme-background-card)",
      },
    },
  } satisfies RehypeShikiOptions)
  .use(rehypeStringify);
```

_* Further adjustments might be needed to retain feature parity if Astro's Markdown parsing configuration changes in the future._
